### PR TITLE
[WPE] Add support for touch based pointer events

### PR DESCRIPTION
**Pull request by Bastian-Krause**
#### 84e4a4ae1e8936da66c30862be411246417d44fd
<pre>
[WPE] Add support for touch based pointer events
<a href="https://bugs.webkit.org/show_bug.cgi?id=214870">https://bugs.webkit.org/show_bug.cgi?id=214870</a>

Reviewed by Carlos Alberto Lopez Perez and Michael Catanzaro.

Currently all non-Apple ports support only pointer events of type
&quot;mouse&quot; because they a mapped 1:1 to mouse events. This commit adds
the pointer event support for touch events for the WPE port. With
this patch it should be easy for other ports to add pointer event support
as well.

The WPE port creates artificial motion, button down, button up events
for short touch gestures in PageClientImpl::doneWithTouchEvent() to keep
compatibility, see:

<a href="https://w3c.github.io/pointerevents/#compatibility-mapping-with-mouse-events">https://w3c.github.io/pointerevents/#compatibility-mapping-with-mouse-events</a>

These NativeWebMouseEvents are undistinguishable from actual mouse
events and lead to additional pointer events of type mouse being
dispatched.
Fix this by passing the SyntheticClickType down to the WebMouseEvent and
PlatformMouseEvent. With this, WPE can make use of the code path
introduced in <a href="https://bugs.webkit.org/show_bug.cgi?id=205551">https://bugs.webkit.org/show_bug.cgi?id=205551</a> due to the
same issue.

Initial patch version by Marco Felsch &lt;m.felsch@pengutronix.de&gt;. Tested
and improved by Bastian Krause &lt;bst@pengutronix.de&gt;.

* Source/WebCore/SourcesWPE.txt: Add dom/wpe/PointerEventWPE.cpp, see
below.
* Source/WebCore/dom/PointerEvent.h: Enable various pointer event paths
for WPE, move buttonForType() and buttonsForType() generics from
PointerEventIOS.
* Source/WebCore/dom/wpe/PointerEventWPE.cpp: Adapted for WPE from
Source/WebCore/dom/ios/PointerEventIOS.cpp.
(WebCore::pointerEventType): Add a translation function similar to
PointerEventIOS&apos;s pointerEventType().
(WebCore::PointerEvent::create): Add create functions similar to
PointerEventIOS.
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTouchEvent): Change point iteration to
make index value accessible. Add pointer target hit tests for
TouchReleased/TouchCancelled/TouchMoved. Implement WPE-specific
workaround to compensate for non-existent touch stationary events.
Finally dispatch pointer events.
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::cancelPointer): Enable existing
pointer event code paths for WPE.
* Source/WebCore/page/PointerCaptureController.h:
(WebCore::PointerCaptureController::CapturingData::hasAnyElement const):
* Source/WebCore/platform/PlatformTouchEvent.h:
(WebCore::PlatformTouchEvent::setTouchPoints): Allow setting touchPoints
for WPE to make the above mentioned touch stationary worarkound
possible.
* Source/WebCore/platform/PlatformTouchPoint.h:
(WebCore::PlatformTouchPoint::PlatformTouchPoint): Allow WPE to create
PlatformTouchPoints to make the above mentioned touch stationary
worarkound possible.
* Source/WebKit/Shared/NativeWebMouseEvent.h: Add synthetic click type.
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent): Pass the
SyntheticClickType to the PlatformMouseEvent, for WPE only because effects
on other ports are unknown at this time.
* Source/WebKit/Shared/libwpe/NativeWebMouseEventLibWPE.cpp:
(WebKit::NativeWebMouseEvent::NativeWebMouseEvent): Add synthetic click
type.
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp: Add synthetic click
type.
(WebKit::WebEventFactory::createWebMouseEvent):
* Source/WebKit/Shared/libwpe/WebEventFactory.h: Pass synthetic click
type to WebMouseEvent.
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent): Set synthetic click type
for synthetic mouse events to NativeWebMouseEvent.

Canonical link: <a href="https://commits.webkit.org/263969@main">https://commits.webkit.org/263969@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/fcef3c978b920c14e6c5a19e6d48056547134a4f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6430 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7806 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6568 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6380 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9448 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6363 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7896 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3836 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5627 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13536 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5701 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6185 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5057 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5597 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1484 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5964 "Built successfully") | | |

### CODE REVIEW

**Comment by Bastian-Krause** [comment]
I am not sure if comments belong here or in the bug ticket, but since the review will happen here, I'll comment here:
I've addressed all previous review feedback, either by integrating it or by commenting on it (see bug ticket).

@clopez Regarding the tests: I tried my best to let the IOS/imported tests run, but no additional test became green. If I understand the situation correctly, the WPEWebKitTestRunner needs to implement various methods to make them actually run. Also note that this PR does not affect touch-action, which seems to be the main focus of most tests.

File: Source/WebCore/dom/PointerEvent.h
Comment (shivamidow) [inline Source/WebCore/dom/PointerEvent.h]: *This is an unofficial review. Just refer.*

For a primitive argument, we can use a variable name in method declarations. `float pressure`

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (shivamidow) [inline Source/WebCore/dom/wpe/PointerEventWPE.cpp]: *This is an unofficial review. Just refer.*

Maybe, what you want to say is.
```
// Where the pressure is not supported, it must be 0.5 in the active button state—otherwise, 0.
// https://www.w3.org/TR/pointerevents/#pointerevent-interface
```

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (shivamidow) [inline Source/WebCore/dom/wpe/PointerEventWPE.cpp]: *This is an unofficial review. Just refer.*

So... by checking these three-pointer events, can we decide if the hardware is capable of detecting pressures?

File: Source/WebCore/page/EventHandler.cpp
Comment (shivamidow) [inline Source/WebCore/page/EventHandler.cpp]: *This is an unofficial review. Just refer.*

touchPointerEvent sounds confusing. handlePointerEventWithTouch()?

File: Source/WebCore/page/EventHandler.cpp
Comment (shivamidow) [inline Source/WebCore/page/EventHandler.cpp]: *This is an unofficial review. Just refer.*

If that is the case, what about putting the most notable change with if-end guards inside handleTouchEvent()?

File: Source/WebCore/page/PointerCaptureController.cpp
Comment (shivamidow) [inline Source/WebCore/page/PointerCaptureController.cpp]: *This is an unofficial review. Just refer.*

#elif PLATFORM(WPE)

File: Source/WebCore/page/PointerCaptureController.h
Comment (shivamidow) [inline Source/WebCore/page/PointerCaptureController.h]: *This is an unofficial review. Just refer.*

bool isPrimary?

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (shivamidow) [inline Source/WebCore/dom/wpe/PointerEventWPE.cpp]: *This is an unofficial review. Just refer.*

The WPE platform uses this file only. So, I think PLATFORM(WPE) is redundant.

File: Source/WebCore/page/PointerCaptureController.cpp
Comment (shivamidow) [inline Source/WebCore/page/PointerCaptureController.cpp]: *This is an unofficial review. Just refer.*

why don't we have an exception handling for the empty type like line 208?

File: Source/WebCore/dom/PointerEvent.h
Comment (Bastian-Krause) [reply Source/WebCore/dom/PointerEvent.h]: Thanks, fixed.

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (Bastian-Krause) [reply Source/WebCore/dom/wpe/PointerEventWPE.cpp]: No, as you can see, we don't pass an actual pressure value here, but rather use 0 and 0.5 as specified for platforms that do not support pressure:
> For hardware and platforms that do not support pressure, the value MUST be 0.5 when in the [active buttons state](https://www.w3.org/TR/pointerevents/#dfn-active-buttons-state) and 0 otherwise. Note: all pointerup events will have pressure 0.

Source: https://www.w3.org/TR/pointerevents/#dom-pointerevent-pressure

Actual pressure handling can be added later. Most touch controllers and drivers do not have this information anyway.

File: Source/WebCore/page/EventHandler.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/EventHandler.cpp]: I'd like to wait for maintainer feedback if this should be renamed.

File: Source/WebCore/page/EventHandler.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/EventHandler.cpp]: I don't think that's a good idea. `handleTouchPointerEvent()` is called from multiple locations and the new function behaves differently in some key aspects. Mixing these will make handleTouchEvent() huge and the if-end guards make it unnecessarily confusing.

File: Source/WebCore/page/PointerCaptureController.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/PointerCaptureController.cpp]: Thanks, fixed.

File: Source/WebCore/page/PointerCaptureController.h
Comment (Bastian-Krause) [reply Source/WebCore/page/PointerCaptureController.h]: Thanks, fixed.

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (Bastian-Krause) [reply Source/WebCore/dom/wpe/PointerEventWPE.cpp]: True, I was confused by the` PLATFORM(IOS)` in the IOS counterpart of this file. I just realized that this is necessary there because only a single `Source/WebCore/SourcesCocoa.txt` exists for Mac/IOS. I've dropped it here.

File: Source/WebCore/page/PointerCaptureController.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/PointerCaptureController.cpp]: The empty type happens only for IOS because `PointerCaptureController::dispatchEventForTouchAtIndex()` is called from there omitting the type argument. All other calls pass a type explicitly.

I've added a `ASSERT(!type.isEmpty())` here.

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (Bastian-Krause) [reply Source/WebCore/dom/wpe/PointerEventWPE.cpp]: No, see above.

**Comment by Bastian-Krause** [comment]
Note: I'll push an updated version of this PR later today.

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (Bastian-Krause) [reply Source/WebCore/dom/wpe/PointerEventWPE.cpp]: This is now reworked completely, please have a look at the recent version.

File: Source/WebCore/dom/wpe/PointerEventWPE.cpp
Comment (Bastian-Krause) [reply Source/WebCore/dom/wpe/PointerEventWPE.cpp]: This has been reworked completely. No need for any of this anymore. Please have a look at the recent version.

File: Source/WebCore/page/EventHandler.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/EventHandler.cpp]: This has been reworked completely. No need for any of this anymore. Please have a look at the recent version.

File: Source/WebCore/page/EventHandler.cpp
Comment (Bastian-Krause) [reply Source/WebCore/page/EventHandler.cpp]: The recent rework made it indeed possible to get rid of the extra function. Please have a look at the recent version.

**Comment by Bastian-Krause** [comment]
Here it is, a complete overhaul. The implementation is now more concise and reuses as much as possible from other platforms. I also tested more corner cases. Seems to work well in the situations I checked. It would be great if someone can have a look at this.

File: Source/WebCore/page/EventHandler.cpp
Comment (clopez) [inline Source/WebCore/page/EventHandler.cpp:4772]: why is this change needed?

Comment (Bastian-Krause) [reply Source/WebCore/page/EventHandler.cpp:4772]: Quote from the commit message:
> (WebCore::EventHandler::handleTouchEvent): Change point iteration to
make index value accessible. [...]

The `index` is used down in the `pointerCaptureController.dispatchEventForTouchAtIndex()` call.

Comment (clopez) [reply Source/WebCore/page/EventHandler.cpp:4772]: I see. I missed that, sorry.

**Comment by Bastian-Krause** [comment]
Ping? Can anyone have a look at this?

**Comment by mcatanzaro** [comment]
It looks like there are a lot of test failures on this one.

**Comment by Bastian-Krause** [comment]
> It looks like there are a lot of test failures on this one.

Any idea how to tackle them? I'll try rebasing first, maybe this can eliminate false-positives.

**Comment by Bastian-Krause** [comment]
@mcatanzaro Looks like rebasing fixed the tests. No idea about the iOS tests, though. I saw other PRs stumbling upon the same error, so I guess it's not related to this PR really.

**Comment by Bastian-Krause** [comment]
@mcatanzaro Would you be willing to pick this one up, too? Or should I ping the WPE folks again?

**Review by mcatanzaro — APPROVED** [review]
I'm not sufficiently familiar with event handling to spot bugs, but this seems generally OK.

So currently only iOS and WPE will support touch pointer events? This means GTK will not, even though it has had touch support for a long time?

**Comment by Bastian-Krause** [comment]
> I'm not sufficiently familiar with event handling to spot bugs, but this seems generally OK.

Cool.

> So currently only iOS and WPE will support touch pointer events?

Correct.

> This means GTK will not, even though it has had touch support for a long time?

Yes. Implementing pointer events of type "touch" for GTK should be trivial with this PR merged, though.

**Comment by Bastian-Krause** [comment]
@clopez Oh, sorry, I must have lost that line during rebase. Fixed.

**Comment by clopez** [comment]
I got a [backtrace from a Debug build](http://sprunge.us/oT7ido) and debugged it a bit.

I think I found the problem.

This happens:

1. At `EventHandler::handleTouchEvent()` in Source/WebCore/page/EventHandler.cpp  `pointState` get assigned a value of `WebCore::PlatformTouchPoint::TouchStationary`
2. But `pointerTarget` doesn't get a value assigned because on the last else we are only assigning a value to  pointerTarget` when `pointState` is  `PlatformTouchPoint::TouchMoved. [See line 4860 at Source/WebCore/page/EventHandler.cpp](https://github.com/WebKit/WebKit/blob/a2dd833816d679576b36b865c0e0f1dc782b1975/Source/WebCore/page/EventHandler.cpp#L4860)
3. We end calling pointerCaptureController().dispatchEventForTouchAtIndex(*pointerTarget) which causes a  crash when trying to dereference a null pointer. [See line 4891 at Source/WebCore/page/EventHandler.cpp](https://github.com/WebKit/WebKit/blob/a2dd833816d679576b36b865c0e0f1dc782b1975/Source/WebCore/page/EventHandler.cpp#L4891)

The following patch fixes the issue and I get multitouch working as expected on the above tests:
```#!patch
diff --git a/Source/WebCore/page/EventHandler.cpp b/Source/WebCore/page/EventHandler.cpp
index ac9ec3e29b01..f6a46c673834 100644
--- a/Source/WebCore/page/EventHandler.cpp
+++ b/Source/WebCore/page/EventHandler.cpp
@@ -4857,10 +4857,8 @@ bool EventHandler::handleTouchEvent(const PlatformTouchEvent& event)
             // No hittest is performed on move or stationary, since the target is not allowed to change anyway.
             touchTarget = m_originatingTouchPointTargets.get(touchPointTargetKey);
 
-            if (pointState == PlatformTouchPoint::TouchMoved) {
-                HitTestResult result = hitTestResultAtPoint(pagePoint, hitType | HitTestRequest::Type::AllowChildFrameContent);
-                pointerTarget = result.targetElement();
-            }
+            HitTestResult result = hitTestResultAtPoint(pagePoint, hitType | HitTestRequest::Type::AllowChildFrameContent);
+            pointerTarget = result.targetElement();
         }
 
         if (!is<Node>(touchTarget))
```

**Review by clopez — CHANGES_REQUESTED** [review]
Please upload a new patch with the above fix or a similar one

**Comment by Bastian-Krause** [comment]
@clopez Thanks for debugging this, appreciate it. I've kept the hit test conditional, because we don't need it for touch stationary events. Instead, `dispatchEventForTouchAtIndex()` is now called only if `pointerTarget` was set previously.

**Review by clopez — APPROVED** [review]

**Pull request merged**
